zio-http/src/main/scala/zio/http/QueryParams.scala

```diff
@@ -169,37 +159,111 @@ final case class QueryParams(map: Map[String, Chunk[String]]) {
   /**
    * Normalizes the query parameters by removing empty keys and values.
    */
-  def normalize: QueryParams =
-    if (isEmpty) self
-    else QueryParams(map.filter(i => i._1.nonEmpty && i._2.nonEmpty))
+  def normalize: QueryParams
 
   /**
    * Removes the specified key from the query parameters.
    */
-  def remove(key: String): QueryParams = QueryParams(map - key)
+  def remove(key: String): QueryParams
 
   /**
    * Removes the specified keys from the query parameters.
    */
-  def removeAll(keys: Iterable[String]): QueryParams = QueryParams(map -- keys)
+  def removeAll(keys: Iterable[String]): QueryParams
 
   /**
    * Converts the query parameters into a form.
    */
   def toForm: Form = Form.fromQueryParams(self)
+
 }
 
-object QueryParams {
+/**
+ * A ordered collection of query parameters.
+ */
+final case class ListMapQueryParams(map: ListMap[String, Chunk[String]]) extends QueryParams {
```

Comment:
I would move this into `QueryParams` companion object and make it package private to zio.http (e.g. `private[http]`). 

Why? 

To avoid polluting the top-level package with something no one will ever use directly, or even know about.

Reply:
Makes sense. Implementation coming.

Reply:
Done